src/algorithms/all_shortest_paths.c

```diff
@@ -0,0 +1,154 @@
+			if(depth > ctx->maxLen || depth >= array_len(ctx->levels)) {
+				// we reached to the max level and didn't found the dest node
+				depth = 0;
+				break;
+			}
+		}
+
+		// get the node from the frontier
+		LevelConnection frontierConnection = array_pop(ctx->levels[depth]);
+		Node frontierNode = frontierConnection.node;
+		NodeID frontierID = ENTITY_GET_ID(&frontierNode);
+		// check if we reached the destination
+		if(destID == frontierID && depth > 0) {
+			break;
+		}
+
+		// the node has already been visited if it is already in the vector
+		bool is_visited;
+		GrB_Info info = GrB_Vector_extractElement_BOOL(&is_visited, visited, frontierID);
+
+		// introduce neighbors only if path depth < maximum path length
+		// and frontier wasn't already expanded
+		if(info == GrB_NO_VALUE && depth < ctx->maxLen) {
+			GrB_Vector_setElement_BOOL(visited, true, frontierID);
+			// add all neighbors of the current node to the next depth level
+			addNeighbors(ctx, &frontierConnection, depth + 1, ctx->dir);
+		}
+	}
+
+	if(depth > 0) {
+		// clean data
+		array_clear(ctx->levels[depth]);
+		depth++; // switch from edge count to node count
+		if(depth < array_len(ctx->levels)) array_clear(ctx->levels[depth]);
+	}
+
+	ctx->visited = visited;
+
+	return depth;
+}
+
+// find paths from src to dest by traversing from dest to src using DFS
+// inspecting nodes which where discovered by
+// the previous call to `AllShortestPaths_FindMinimumLength`
+Path *AllShortestPaths_NextPath
+(
+	AllPathsCtx *ctx
+) {
+	ASSERT(ctx != NULL);
+
+	// TODO: add an explanation on how paths are discovered
+
+	uint32_t depth = Path_NodeCount(ctx->path);
+	if(depth > 0) {
+		// a full path already returned
+		// reverse it to be from dest to src
+		// pop the src node and edge
+		// advance to next path by backtracking
+		Path_Reverse(ctx->path);
+		Path_PopNode(ctx->path);
+		Path_PopEdge(ctx->path);
+		depth--;
+	} else {
+		if(array_len(ctx->levels[0]) == 0) return NULL;
+
+		LevelConnection frontierConnection = array_pop(ctx->levels[0]);
+		Node frontierNode = frontierConnection.node;
+		Path_AppendNode(ctx->path, frontierNode);
+		depth++;
+		addNeighbors(ctx, &frontierConnection, depth, ctx->dir);
+	}
+
+	// as long as we didn't found a full path from src to dest
+	while (depth < ctx->maxLen) {
+		if (array_len(ctx->levels[depth]) > 0) {
+			// get a new node from the frontier
+			bool is_visited;
+			LevelConnection frontierConnection = array_pop(ctx->levels[depth]);
+			Node frontierNode = frontierConnection.node;
+			NodeID frontierID = ENTITY_GET_ID(&frontierNode);
+			GrB_Info info = GrB_Vector_extractElement_BOOL(&is_visited,
+					ctx->visited, frontierID);
+
+			// consider only on visited nodes
+			if(info == GrB_NO_VALUE) continue;
+
+			// if we reached to the end of the path and this node is not the
+			// dst node continue
+			if(depth == ctx->maxLen - 1 && frontierID != ENTITY_GET_ID(ctx->dst)) continue;
+
+			Path_AppendNode(ctx->path, frontierNode);
```

Comment:
line 108 ` Path_AppendNode(ctx->path, frontierNode);` pushes the first node to the path, after which line 131 will eventually be called pushing a second node to the path, so now we have 2 consecutive nodes on the path instead of alternating node edge node edge... sequence, or is it that I'm missing something ?

Reply:
Path has 2 arrays one for nodes and one for edge